crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -535,6 +572,81 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_nodes_disruptive(to_upgrade)
+        to_upgrade.each do |element|
+          # If role has a cluster assigned ugprade that cluster (reuse the non-disruptive code here)
+          if ServiceObject.is_cluster?(element)
+            cluster = ServiceObject.cluster_name element
+            cluster_env = "pacemaker-config-#{cluster}"
+            founder = ::Node.find(
+              "pacemaker_founder:true AND " \
+              "pacemaker_config_environment:#{cluster_env}"
+            ).first
+            upgrade_cluster founder, cluster_env
+          else
+            # if role has single node(s) assigned upgrade those nodes (optionally all at once)
+            # FIXME: could we do it in paralel?
+            upgrade_one_node element
+          end
+        end
+
+        # and handle those as part of the compute node upgrade
+      end
+
+      # Go through active proposals and return elements with assigned roles in the right order
+      def upgradable_elements_of_proposals(proposals)
```

Comment:
In general this function seems to have an issue with mixed cluster/non-cluster setups. E.g. if keystone is deployed non-HA on a node that is a cluster member for some later barclamp (glance, neutron, cinder, ...) it won't upgrade the whole cluster.

I know this is not really a supported setup (and no sane person would install that it that way), but we might want to deal with it in  the disruptive upgrade path somehow. (Solution would be to load the node objects and check if they are a cluster member).